File: 剑指Offer/29.顺时针打印矩阵.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0) return vector<int>();
        int rows = matrix.size(), cols = matrix[0].size();
        int topRow = 0, leftCol = 0, bottomRow = rows - 1, rightCol = cols - 1;
        int counts = rows * cols;
        vector<int> res(counts);
        for (int k = 0; k < counts;) {
            for (int i = leftCol; i <= rightCol; i++) {
                res[k++] = matrix[topRow][i];
            }
            topRow++;
            if (k >= counts) break;
            for (int i = topRow; i <= bottomRow; i++) {
                res[k++] = matrix[i][rightCol];
            }
            rightCol--;
            if (k >= counts) break;
            for (int i = rightCol; i >= leftCol; i--) {
                res[k++] = matrix[bottomRow][i];
            }
            bottomRow--;
            if (k >= counts) break;
            for (int i = bottomRow; i >= topRow; i--) {
                res[k++] = matrix[i][leftCol];
            }
            leftCol++;
        }
        return res;
    }
};

File: 剑指Offer/29.顺时针打印矩阵.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return []

        topRow = 0
        bottomRow = len(matrix) - 1
        leftCol = 0
        rightCol = len(matrix[0]) - 1

        counts = (bottomRow + 1) * (rightCol + 1)
        res = []

        while counts > 0:
            for i in range(leftCol, rightCol + 1):
                res.append(matrix[topRow][i])
                counts -= 1
            topRow += 1
            if counts <= 0: break

            for i in range(topRow, bottomRow + 1):
                res.append(matrix[i][rightCol])
                counts -= 1
            rightCol -= 1
            if counts <= 0: break

            for i in range(rightCol, leftCol - 1, -1):
                res.append(matrix[bottomRow][i])
                counts -= 1
            bottomRow -= 1
            if counts <= 0: break

            for i in range(bottomRow, topRow - 1, -1):
                res.append(matrix[i][leftCol])
                counts -= 1
            leftCol += 1
        return res